src/mono/wasm/runtime/exports-binding.ts

```diff
@@ -0,0 +1,164 @@
+    mono_wasm_fire_debugger_agent_message_with_data,
+    mono_wasm_fire_debugger_agent_message_with_data_to_pause,
+    // mono-threads-wasm.c
+    schedule_background_exec,
+
+    // interp.c and jiterpreter.c
+    mono_interp_tier_prepare_jiterpreter,
+    mono_interp_record_interp_entry,
+    mono_interp_jit_wasm_entry_trampoline,
+    mono_interp_jit_wasm_jit_call_trampoline,
+    mono_interp_invoke_wasm_jit_call_trampoline,
+    mono_interp_flush_jitcall_queue,
+    mono_jiterp_do_jit_call_indirect,
+
+    mono_wasm_profiler_enter,
+    mono_wasm_profiler_leave,
+
+    // driver.c
+    mono_wasm_trace_logger,
+    mono_wasm_set_entrypoint_breakpoint,
+    mono_wasm_event_pipe_early_startup_callback,
+
+    // corebindings.c
+    mono_wasm_release_cs_owned_object,
+    mono_wasm_bind_js_function,
+    mono_wasm_invoke_bound_function,
+    mono_wasm_invoke_import,
+    mono_wasm_bind_cs_function,
+    mono_wasm_marshal_promise,
+    mono_wasm_change_case_invariant,
+    mono_wasm_change_case,
+    mono_wasm_compare_string,
+    mono_wasm_starts_with,
+    mono_wasm_ends_with,
+    mono_wasm_index_of,
+    mono_wasm_get_calendar_info,
+    mono_wasm_get_culture_info,
+    mono_wasm_get_first_day_of_week,
+    mono_wasm_get_first_week_of_year,
+];
+
+const wasmImports: Function[] = [
+    ...mono_wasm_imports,
+    // threading exports, if threading is enabled
+    ...mono_wasm_threads_imports,
+    // legacy interop exports, if enabled
+    ...mono_wasm_legacy_interop_imports
+];
+
+export function replace_linker_placeholders(imports: WebAssembly.Imports) {
+    // the output from emcc contains wrappers for these linker imports which add overhead,
+    //  but now we have what we need to replace them with the actual functions
+    // By default the imports all live inside of 'env', but emscripten minification could rename it to 'a'.
+    // See https://github.com/emscripten-core/emscripten/blob/c5d1a856592b788619be11bbdc1dd119dec4e24c/src/preamble.js#L933-L936
+    const env = imports.env || imports.a;
+    if (!env) {
+        mono_log_warn("WARNING: Neither imports.env or imports.a were present when instantiating the wasm module. This likely indicates an emscripten configuration issue.");
+        return;
+    }
+
+    // the import names could be minified by applyImportAndExportNameChanges in emcc
+    // but the stubs are not, so we can use the stub names to find the short names
+    const indexToNameMap: string[] = new Array(wasmImports.length);
+    for (const shortName in env) {
+        const fn = env[shortName] as Function;
+        if (typeof fn === "function" && fn.toString().indexOf("runtime_idx") !== -1) {
+            try {
+                const { runtime_idx } = fn();
+                indexToNameMap[runtime_idx] = shortName;
+            } catch {
+                // no-action
+            }
+        }
+
+        let idx = 0;
+        for (const fn of wasmImports) {
+            const shortName = indexToNameMap[idx];
```

Comment:
Shadowing the outer `shortName in env` with this `shortName` is probably not a good idea

Reply:
oh, it should not be nested, thanks!